webhookValidator.go

```diff
@@ -0,0 +1,189 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package ancla
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+)
+
+var (
+	errInvalidURL            = errors.New("Invalid Config URL")
+	errInvalidFailureURL     = errors.New("Invalid Failure URL")
+	errInvalidAlternativeURL = errors.New("Invalid Alternative URL(s)")
+	errURLIsNotHTTPS         = errors.New("URL is not HTTPS")
+	errInvalidHost           = errors.New("Invalid host")
+	errIPGivenAsHost         = errors.New("Cannot use IP as host")
+	errLocalHostGivenAsHost  = errors.New("Cannot use Localhost as host")
+	errLoopbackGivenAsHost   = errors.New("Cannot use loopback host")
+)
+
+type Validator interface {
+	Validate(w Webhook) error
+}
+
+type Validators []Validator
+
+type ValidFunc func(Webhook) error
+
+type ValidURLFunc func(*url.URL) error
+
+// Validate runs the given webhook through each validator in the validators list.
+// It returns as soon as the webhook is considered invalid and returns nil if the
+// webhook is valid.
+func (vs Validators) Validate(w Webhook) error {
+	for _, v := range vs {
+		err := v.Validate(w)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GoodURL parses the given webhook's Config.URL, FailureURL, and Config.AlternativeURLs
+// and returns as soon as the URL is considered invalid. It returns nil if the URL is
+// valid.
+func GoodURL(v []ValidURLFunc) ValidFunc {
+	return func(w Webhook) error {
+		//check Config.URL
+		parsedURL, err := url.ParseRequestURI(w.Config.URL)
+		if err != nil {
+			return fmt.Errorf("%w: %v", errInvalidURL, err)
+		}
+		for _, f := range v {
+			if f == nil {
+				continue
+			}
+			err = f(parsedURL)
+			if err != nil {
+				return fmt.Errorf("%w: %v", errInvalidURL, err)
+			}
+		}
+		//check FailureURL if it exists
+		if w.FailureURL != "" {
+			parsedFailureURL, err := url.ParseRequestURI(w.FailureURL)
+			if err != nil {
+				return fmt.Errorf("%w: %v", errInvalidFailureURL, err)
+			}
+			for _, f := range v {
+				if f == nil {
+					continue
+				}
+				err = f(parsedFailureURL)
+				if err != nil {
+					return fmt.Errorf("%w: %v", errInvalidFailureURL, err)
+				}
+			}
+		}
+		//check AlternativeURLs if they exist
+		if len(w.Config.AlternativeURLs) > 0 {
+			for _, u := range w.Config.AlternativeURLs {
+				if u != "" {
+					parsedAlternativeURL, err := url.ParseRequestURI(u)
+					if err != nil {
+						return fmt.Errorf("%w: %v", errInvalidAlternativeURL, err)
+					}
+					for _, f := range v {
+						if f == nil {
+							continue
+						}
+						err = f(parsedAlternativeURL)
+						if err != nil {
+							return fmt.Errorf("%w: %v", errInvalidAlternativeURL, err)
+						}
+					}
+				}
+			}
+		}
+		return nil
+	}
+}
+
+// Validate runs the function and returns the result.
```

Comment:
```suggestion
// Validate runs the function and returns the result.  This allows any ValidFunc to implement the Validator interface.
```

Reply:
This could probably be split into two lines, I'm just not sure where since that's easier to figure out in an IDE.